### ISSUE TITLE
[NEW] Support connection/authentication with URLs that has subdirectories

### ISSUE THREAD

**Issue by Alacard** [RocketChat/Rocket.Chat.Android#617]
- Your Rocket.Chat.Android app version: ####
Rocket.Chat+ 1.0.32
- Your Rocket.Chat server version: ####
0.59.6
- Device model (or emulator) you're running with: ####
Nexus 6P - Android 7.1.2
Nexus 5X - Android 7.1.2
Amazon Fire Tablet 8.9 - Android 4.4.4
Alcatel - Android 6.0
- Steps to reproduce:
Install Rocket.Chat server with reverse proxy w/SSL w/sub directory

What i found out is that if you have a sub directory setup the app bugs out. so bad that the connection attempt dont even happen when you try to login. (it dont crash just have the infinate loading icon) what i tryed was this.

Server =works
Server + SSL=works
Server + reverse proxy = works
Server + reverse proxy + SSL= works
Server + subdirectory=dont works
Server + SSL + sub directory= dont work
Server + reverse proxy + sub directory = dont work
Server + reverse proxy + SSL + sub directory = dont work

from using wireshark and apache logs no connection is even tempted when using a sub directory. 

as for my reverse proxy config this what i have

```
LoadModule proxy_module modules/mod_proxy.so
LoadModule proxy_http_module modules/mod_proxy_http.so
LoadModule proxy_wstunnel_module modules/mod_proxy_wstunnel.so
LoadModule rewrite_module modules/mod_rewrite.so

<Location /rocketchat>
    Order allow,deny
    Allow from all

    RewriteEngine On
	RewriteCond %{HTTP:Upgrade} =websocket [NC]
	RewriteRule /share/Web/(.*) ws://192.168.1.117:3000/$1 [P,L]
	RewriteCond %{HTTP:Upgrade} !=websocket [NC]
	RewriteRule /share/Web/(.*) http://192.168.1.117:3000/$1 [P,L]
	
ProxyPassReverse http://192.168.1.117:3000/
</Location>
```

dont think its my config seeing as still have a problem even if reverse proxy was not being used only sub directory

P.S. the url im trying to use is https://subdomain.domain.com/rocketchat
P.S.S that url works fine for bowsers (firefox,chrome,Android chrome) even the windows client Rocket.chat+

Edit: Just seen theres a unreleased version of the app on google play called "Rocket.Chat.ReactNative (Unreleased)" when entering the url of the server it just says "is not a valid Rocket.Chat Instance" cant even get to the login screen on that one.

Edit 2: almost forgot
ROOT_URL=https://subdomain.domain.com/rocketchat
Site URL (under settings)=https://subdomain.domain.com/rocketchat

**Comment by Alacard** [comment]
finally got wireshark to log it, seems like i was right its not obeying the url, its not trying to connect to https://subdomain.domain.com/rocketchat/websocket is trying to connect to https://subdomain.domain.com/websocket. seeing as https://subdomain.domain.com/ dont redirect to the server its getting a 404

```
GET /websocket HTTP/1.1
origin: subdomain.domain.com
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Key: ei1Jijh5T/HKiXtaMagQiQ==
Sec-WebSocket-Version: 13
Host: subdomain.domain.com
Accept-Encoding: gzip
User-Agent: okhttp/3.6.0

HTTP/1.1 404 Not Found
Date: Tue, 05 Dec 2017 07:44:40 GMT
Server: http server 1.0
X-Frame-Options: SAMEORIGIN
Content-type: text/html
Last-modified: Tue, 05 Dec 2017 07:44:40 GMT
Accept-Ranges: bytes
Transfer-Encoding: chunked
```

**Comment by rafaelks** [comment]
@Alacard Thanks for the report! This is a known issue, I'll let your issue opened until we solve this. Is it possible to you to don't use subdirectory right now? Only subdomains are being supported.

**Comment by Alacard** [comment]
Wife work wifi connection only allows 80 and 443 ports everything else is blocked, I have other services running on 80 and 443 so cant use them, so subdirectory is the only option.  seeing as this the only one that got the wife factor approval (she like the layout) ill just have to wait (know the hard part convincing her to wait it out XD /cry)

Edit: just thinking out loud, what would happen if i was to reverse proxy /websocket to rocketchat/websocket. wonder if that would work. ill have to try that when i get home from work, would be a nice work around

**Comment by Alacard** [comment]
what able to get it to finally connect by using this config

```
<Location /rocketchat>
    Order allow,deny
    Allow from all
	
    RewriteEngine On
	RewriteCond %{HTTP:Upgrade} =websocket [NC]
	RewriteRule /share/Web/(.*) ws://192.168.1.117:3000/$1 [P,L]
	RewriteCond %{HTTP:Upgrade} !=websocket [NC]
	RewriteRule /share/Web/(.*) http://192.168.1.117:3000/$1 [P,L]
	
ProxyPassReverse http://192.168.1.117:3000/
</Location>

<Location /websocket>
    Order allow,deny
    Allow from all
	
    RewriteEngine On
	RewriteCond %{HTTP:Upgrade} =websocket [NC]
	RewriteRule /share/Web/(.*) ws://192.168.1.117:3000/rocketchat/$1 [P,L]
	RewriteCond %{HTTP:Upgrade} !=websocket [NC]
	RewriteRule /share/Web/(.*) http://192.168.1.117:3000/rocketchat/$1 [P,L]
	
ProxyPassReverse ws://192.168.1.117:3000/rocketchat/
</Location>

<Location /sockjs>
    Order allow,deny
    Allow from all
	
    RewriteEngine On
	RewriteCond %{HTTP:Upgrade} =websocket [NC]
	RewriteRule /share/Web/(.*) ws://192.168.1.117:3000/rocketchat/$1 [P,L]
	RewriteCond %{HTTP:Upgrade} !=websocket [NC]
	RewriteRule /share/Web/(.*) http://192.168.1.117:3000/rocketchat/$1 [P,L]
	
ProxyPassReverse ws://192.168.1.117:3000/rocketchat/
</Location>
```

but ran in to another problem, that the images dont load i can redirect that directory but i got a feeling that im just heading down a rabbit hole, thats not going to end good... 

welp back to work on lunch ;_;

**Comment by rafaelks** [comment]
@Alacard For images to load, make sure that your setting `Sire_Url` is correct in your admin panel, please.

**Comment by rafaelks** [comment]
Closing now, because it's already implemented in [2.0.0](https://github.com/RocketChat/Rocket.Chat.Android/milestone/6).
